File: src/main/java/org/gangel/graphomance/metrics/Metrics.java
package org.gangel.graphomance.metrics;

import com.codahale.metrics.MetricRegistry;
import com.codahale.metrics.SharedMetricRegistries;
import com.codahale.metrics.Timer;

import java.util.function.Supplier;

public class Metrics {

    private Metrics() {
    }

    public static Timer GET_NODE = new Timer();
    public static Timer GET_RELATION = new Timer();
    public static Timer CREATE_NODE = new Timer();
    public static Timer CREATE_RELATION = new Timer();
    public static Timer UPDATE_NODE = new Timer();
    public static Timer UPDATE_RELATION = new Timer();

    public static Timer FIND_SHORTEST_PATH = new Timer();

    public static void init() {
        MetricRegistry registry = new MetricRegistry();
        registry.register("CreateNode", CREATE_NODE);
        registry.register("CreateRelation", CREATE_RELATION);
        registry.register("GetNode", GET_NODE);
        registry.register("GetRelation", GET_RELATION);
        registry.register("UpdateNode", UPDATE_NODE);
        registry.register("UpdateRelation", UPDATE_RELATION);
        registry.register("FindShortestPath", FIND_SHORTEST_PATH);
        SharedMetricRegistries.setDefault("default", registry);
    }

    public static <T> T measure(Timer metric, Supplier<T> body) {
        try(Timer.Context ts = metric.time()) {
            return body.get();
        }
    }

    public static void measure(Timer metric, Runnable body) {
        try(Timer.Context ts = metric.time()) {
            body.run();
        }
    }

}

File: src/main/kotlin/org/graphomance/metrics/Metrics.kt
package org.graphomance.metrics

import com.codahale.metrics.MetricRegistry
import com.codahale.metrics.SharedMetricRegistries
import com.codahale.metrics.Timer
import java.util.function.Supplier

object Metrics {
    var GET_NODE = Timer()
    var GET_RELATION = Timer()
    var CREATE_NODE = Timer()
    var CREATE_RELATION = Timer()
    var UPDATE_NODE = Timer()
    var UPDATE_RELATION = Timer()
    var DELETE_NODE = Timer()
    val DELETE_RELATIONSHIP = Timer()
    var FIND_SHORTEST_PATH = Timer()
    fun init() {
        val registry = MetricRegistry()
        registry.register("CreateNode", CREATE_NODE)
        registry.register(
            "CreateRelation",
            CREATE_RELATION
        )
        registry.register("GetNode", GET_NODE)
        registry.register(
            "GetRelation",
            GET_RELATION
        )
        registry.register("UpdateNode", UPDATE_NODE)
        registry.register(
            "UpdateRelation",
            UPDATE_RELATION
        )
        registry.register(
            "FindShortestPath",
            FIND_SHORTEST_PATH
        )
        registry.register("DeleteNode", DELETE_NODE)
        registry.register("DeleteRelationship", DELETE_RELATIONSHIP)
        SharedMetricRegistries.setDefault("default", registry)
    }

    fun <T> measure(metric: Timer?, body: Supplier<T>): T {
        metric!!.time()
            .use { ts -> return body.get() }
    }

    fun measure(metric: Timer?, body: Runnable) {
        metric!!.time()
            .use { ts -> body.run() }
    }
}